akka-stream/src/main/scala/akka/stream/remote/scaladsl/SinkRef.scala

```diff
@@ -0,0 +1,274 @@
+/**
+ * Copyright (C) 2014-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.remote.scaladsl
+
+import java.util.Queue
+
+import akka.actor.{ ActorRef, Terminated }
+import akka.event.Logging
+import akka.stream._
+import akka.stream.actor.{ MaxInFlightRequestStrategy, RequestStrategy, WatermarkRequestStrategy }
+import akka.stream.impl.FixedSizeBuffer
+import akka.stream.remote.StreamRefs
+import akka.stream.remote.impl.StreamRefsMaster
+import akka.stream.scaladsl.Source
+import akka.stream.stage._
+import akka.util.ByteString
+
+import scala.concurrent.{ Future, Promise }
+
+object SinkRef {
+  def source[T](): Source[T, Future[SinkRef[T]]] =
+    Source.fromGraph(new SinkRefTargetSource[T]()) // TODO settings?
+
+  def bulkTransferSource(port: Int = -1): Source[ByteString, SinkRef[ByteString]] = {
+    ???
+  }
+}
+
+/**
+ * This stage can only handle a single "sender" (it does not merge values);
+ * The first that pushes is assumed the one we are to trust
+ */
+final class SinkRefTargetSource[T]() extends GraphStageWithMaterializedValue[SourceShape[T], Future[SinkRef[T]]] {
+  val out: Outlet[T] = Outlet[T](s"${Logging.simpleName(getClass)}.out")
+  override def shape = SourceShape.of(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val promise = Promise[SinkRef[T]]()
+
+    val logic = new TimerGraphStageLogic(shape) with StageLogging with OutHandler {
+      private[this] lazy val streamRefsMaster = StreamRefsMaster(ActorMaterializerHelper.downcast(materializer).system)
+      private[this] lazy val settings = streamRefsMaster.settings
+
+      private[this] var self: GraphStageLogic.StageActor = _
+      private[this] lazy val selfActorName = streamRefsMaster.nextSinkRefTargetSourceName()
+      private[this] implicit def selfSender: ActorRef = self.ref
+
+      // demand management ---
+      private val highDemandWatermark = 16
+
+      private var expectingSeqNr: Long = 0L
+      private var localCumulativeDemand: Long = 0L // initialized in preStart with settings.initialDemand
+
+      private val receiveBuffer = FixedSizeBuffer[T](highDemandWatermark)
```

Comment:
TODO, write some tests that expose the need for this; i.e. a slow reader from this while we requested many elements from remote.